Overhaul the `/departments/<department_id>` view

abandoned-prototype opened the pull request:
## Description of Changes

I ran into this when working on #1148

The BrowseForm was basically only used as a data-container, this change is making sure we also use it to filter the incoming data to simplify the code in the view file. I also included other improvements like reducing the number of SQL queries sent.

Functionality wasn't changed (at least that was the goal), so the user experience should remain the same.


## Notes for Deployment
N/A

## Screenshots (if appropriate)


## Tests and Linting
- [x] This branch is up-to-date with the `develop` branch.
- [ ] Ran `make create_db_diagram` command.
- [x] `pytest` passes on my local development environment.
- [x] `pre-commit` passes on my local development environment.

abandoned-prototype committed:
Overhaul the /departments/<department_id> view.

This is making better use of the BrowseForm for filtering to avoid a lot of boilerplate code and adds some other improvements like sending fewer sql queries (by avoiding some N+1 issues).

abandoned-prototype commented on OpenOversight/app/main/views.py:
```diff
@@ -866,173 +867,69 @@ def edit_department(department_id: int):
 @main.route("/departments/<int:department_id>")
 def list_officer(
     department_id: int,
-    page: int = 1,
-    race=None,
-    gender=None,
-    rank=None,
-    min_age=None,
-    max_age=None,
-    last_name=None,
-    first_name=None,
-    badge=None,
-    unique_internal_identifier=None,
-    unit=None,
-    current_job=None,
-    require_photo: Optional[bool] = None,
```
flask was never setting these vars, so it is basically just a variable initializer list

abandoned-prototype commented on OpenOversight/app/main/views.py:
```diff
@@ -866,173 +867,69 @@ def edit_department(department_id: int):
-        ),
+        url_for("main.list_officer", department_id=department_id)
+        + "?"
+        + urlencode(request.args.to_dict(flat=False), doseq=True),
```
This doesn't look the cleanest, but unfortunately flask's `url_for` does not have a keyword to supply the get arguments separately and this works.

sea-kelp replied:
This seems to work for me:
```python
url_for("main.list_officer", department_id=department_id, **request.args),
```

Would you mind giving that a try?

abandoned-prototype replied:
This is what I started with, but if someone puts `department_id=` in the url it will cause a server error which I wanted to avoid

michplunkett commented on OpenOversight/app/utils/general.py:
```diff
@@ -159,6 +159,11 @@ def str_is_true(str_) -> bool:
     return bool(strtobool(str_.lower()))
 
 
+def url_for_target_null_values_skipped(target, **kwargs):
```
Could you put types (parameters and returns) on all of these functions?

abandoned-prototype replied:
Done

sea-kelp commented on OpenOversight/app/main/views.py:
```diff
@@ -866,173 +867,69 @@ def edit_department(department_id: int):
-        "rank": [(rc, rc) for rc in rank_selections],
-        "unit": [(uc, uc) for uc in unit_selections],
-    }
-
-    next_url = url_for(
+        if officer.face:
+            face_image_id = sorted(
+                officer.face, key=lambda x: x.featured, reverse=True
+            )[0].img_id
+            img_id_to_officer[face_image_id] = officer
+    images = Image.query.filter(Image.id.in_(img_id_to_officer.keys())).all()
+    for image in images:
+        img_id_to_officer[image.id].image = serve_image(image.filepath)
```
On my end, I'm seeing that many officers who have images available are showing up as "No Photos" in the browse view. This is happening because multiple officers have the same image assigned to them.

I'm not sure that this happens in practice since we usually use face thumbnails but it might be an edge case want to account for.

abandoned-prototype replied:
Yeah, that shouldn't happen in practice, but I agree with the concern and will make adjustments

abandoned-prototype replied:
Fixed this, based on my tests the issue should be resolved

sea-kelp commented on OpenOversight/app/templates/list_officer.html:
```diff
@@ -91,11 +91,9 @@ <h3 class="panel-title accordion-toggle">Officer Rank, Unit, and Status</h3>
                                 name="rank"
                                 class="select2-multi-search"
                                 multiple="multiple">
-                          {% for rank_key, rank in choices['rank'] %}
```
For awareness, a future version of wtforms may break this:
https://github.com/pallets-eco/wtforms/issues/811

abandoned-prototype replied:
Good call! I will change it so it works for both versions

abandoned-prototype replied:
Made that change, it should now also work in the future versions

abandoned-prototype committed:
Fix remaining issues.

Making the template futureproof to wtform change.
Also prevent issues if two officers have the same image set (which should not happen in production I think).
